sklego/preprocessing.py

```diff
@@ -9,6 +9,94 @@
 from sklego.common import TrainOnlyTransformerMixin, as_list
 
 
+def _mk_average(xs, ys, intervals, span=1, method="average"):
+    """
+    asdafa
+    Creates smoothed averages of `ys` at the intervals given by `intervals`.
+    :param xs: all the datapoints of a feature (represents the x-axis)
+    :param ys: all the datapoints what we'd like to predict (represents the y-axis)
+    :param intervals: the intervals at which we'd like to get a good average value
+    :param span: if the method is `average` then this is the span around the interval
+    that is used to determine the average `y`-value, if the method is `normal` the span
+    becomes the value of sigma that is used for weighted averaging
+    :param method: the method that is used for smoothing, can be either `average` or `normal`.
+    :return:
+        An array as long as `intervals` that represents the average `y`-values at those intervals.
+    """
+    if method == "average":
+        distances = 1 / (0.01 + np.abs(xs - intervals))
+        predicate = (xs < (intervals + span)) | (xs < (intervals - span))
+    elif method == "normal":
+        distances = np.exp(-((xs - intervals) ** 2) / span)
+        predicate = xs == xs
+    else:
+        raise ValueError("method needs to be either `average` or `normal`")
+    subset = ys[predicate]
+    dist_subset = distances[predicate]
+    return np.average(subset, weights=dist_subset)
+
+
+class IntervalEncoder(TransformerMixin, BaseEstimator):
+    """
+    The interval encoder bends features in `X` with regards to`y`.
+    We take each column in X separately and smooth it towards `y` using
+    the strategy that is defined in `method`.
+    Note that this allows us to make certain features strictly monotonic
+    in your machine learning model if you follow this with an appropriate
+    model.
+    :param n_chunks: the number of cuts that makes the interval
+    :param method: the interpolation method used
+    :param span: a hyperparameter for the interpolation method, if the
+    method is `normal` it resembles the width of the radial basis
+    function used to weigh the points
+    """
+
+    def __init__(self, n_chunks=10, span=1, method="normal"):
+        allowed_methods = ["average", "normal"]
+        if method not in allowed_methods:
+            raise ValueError(f"`method` must be in {allowed_methods}, got `{method}`")
+        self.span = span
+        self.method = method
+        self.n_chunks = n_chunks
+
+    def fit(self, X, y):
+        """Fits the estimator"""
+        # these two matrices will have shape (columns, quantiles)
+        # quantiles indicate where the interval split occurs
```

Comment:
Do we need additional non-negativity validation on span and n_chunks?

Reply:
That feature is not built yet. Once there, yes.